File: src/com/easy/IsomorphicStrings_205.java
/**
 * 
 */
package com.easy;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class IsomorphicStrings_205 {

    // Array count
    // Time Complexity: O(n)
    // Space Complexity: O(2n) = O(n)
    // Runtime: 4 ms, faster than 87.99%
    // Memory Usage: 34.9 MB, less than 87.92%
    public boolean isIsomorphic(String s, String t) {
        int[] mFirstStrCount = new int[256];
        int[] mSecondStrCount = new int[256];
        int mLength = s.length();

        for (int i = 0; i < mLength; i++) {
            mFirstStrCount[s.charAt(i)] += (i + 1);
            mSecondStrCount[t.charAt(i)] += (i + 1);
            if (mFirstStrCount[s.charAt(i)] != mSecondStrCount[t.charAt(i)]) {
                return false;
            }
        }

        return true;

    }
}

File: Golang_src/easy/isomorphicStrings_205.go
package easy

// Use Array count:
// Time Complexity: O(n)
// Space Complexity:O(1)
// Runtime: 4 ms, faster than 58.49%
// Memory Usage: 3.4 MB, less than 14.15%
func isIsomorphic(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	countS, countT := [256]int{}, [256]int{}

	for i := range s {
		if countS[s[i]] != countT[t[i]] {
			return false
		}
		countS[s[i]] = (i + 1)
		countT[t[i]] = (i + 1)

	}

	return true
}
